### ISSUE TITLE
words fall on other words

### ISSUE THREAD

**Issue by Keramatfar** [amueller/word_cloud#488]
I want to make a word cloud from phrases separated using , instead of space. So after some search i found the following code that apparently make word cloud from a dictionary of words. So i can just replace type of split to what i want and the remaining of code is the same. But when tried the code in the default shape i encountered with the problem that some words have intersection with each other, despite that in standard form usage of library that works quite well.

`import multidict as multidict

import numpy as np

import os
import re
from PIL import Image
from os import path
from wordcloud import WordCloud
import matplotlib.pyplot as plt

def getFrequencyDictForText(sentence):
    fullTermsDict = multidict.MultiDict()
    tmpDict = {}

    # making dict for counting frequencies
    for text in sentence.split(" "):
        if re.match("a|the|an|the|to|in|for|of|or|by|with|is|on|that|be", text):
            continue
        val = tmpDict.get(text, 0)
        tmpDict[text.lower()] = val + 1
    for key in tmpDict:
        fullTermsDict.add(key, tmpDict[key])
    return fullTermsDict

def makeImage(text):
    alice_mask = np.array(Image.open("alice_mask.png"))

    wc = WordCloud(background_color="white", max_words=1000, mask=alice_mask)
    # generate word cloud
    wc.generate_from_frequencies(text)

    # show
    plt.imshow(wc, interpolation="bilinear")
    plt.axis("off")
    plt.show()

# get data directory (using getcwd() is needed to support running example in generated IPython notebook)
d = path.dirname(__file__) if "__file__" in locals() else os.getcwd()

text = open(path.join(d, 'alice.txt'), encoding='utf-8')
text = text.read()
makeImage(getFrequencyDictForText(text))`

**Comment by amueller** [comment]
Can you please provide a fully self-contained, ideally minimal example?
And please post the output image as well.

**Comment by Keramatfar** [comment]
Actually this code does not work:

import multidict as multidict

import numpy as np

import os
import re
from PIL import Image
from os import path
from wordcloud import WordCloud
import matplotlib.pyplot as plt


def getFrequencyDictForText(sentence):
    fullTermsDict = multidict.MultiDict()
    tmpDict = {}

    # making dict for counting frequencies
    for text in sentence.split(","):
        if re.match("a|the|an|the|to|in|for|of|or|by|with|is|on|that|be", text):
            continue
        val = tmpDict.get(text, 0)
        tmpDict[text.lower()] = val + 1
    for key in tmpDict:
        fullTermsDict.add(key, tmpDict[key])
    return fullTermsDict


def makeImage(text):
    #alice_mask = np.array(Image.open("alice_mask.png"))

    wc = WordCloud(background_color="white", max_words=1000)
    # generate word cloud
    wc.generate_from_frequencies(text)

    # show
    plt.imshow(wc, interpolation="bilinear")
    plt.axis("off")
    plt.show()


# get data directory (using getcwd() is needed to support running example in generated IPython notebook)
d = path.dirname(__file__) if "__file__" in locals() else os.getcwd()

text = open(path.join(d, 'text.txt'), encoding='utf-8')
text = text.read()
makeImage(getFrequencyDictForText(text))

![image](https://user-images.githubusercontent.com/16587840/59245311-c0464080-8c07-11e9-8bab-64d47ed2d7e9.png)

**Comment by Keramatfar** [comment]
Dear amueller,
Could not find a solution?

**Comment by henriquebortoletti** [comment]
Hi guys, I think that the issue is in this part of the code
`        val = tmpDict.get(text, 0)
        tmpDict[text.lower()] = val + 1`
The val could take zero even for a word being in your data, because when you put the words in the map, you are putting the words in lower case, but when you are searching for the value of the word, you are searching for the word not being lower case. 
The correct code could be something like this:
`        val = tmpDict.get(text.lower(), 0)
        tmpDict[text.lower()] = val + 1`

**Comment by amueller** [comment]
@Keramatfar sorry for the slow reply. Your example is not self-contained because it doesn't include text.txt so I can't run it to reproduce.

I'm not sure what it is you're trying to accomplish which makes it hard for me to help you.

**Comment by amueller** [comment]
@henriquebortoletti WordCloud should work for any dictionary you provide to ``generate_from_frequencies``. If it doesn't, it would be great if you could provide an example to reproduce the issue.

If you just want to replace the space in phrases by underscores, this would do it:
```python
wordcloud = WordCloud().generate(text)
underscore_separated = {k.replace(" ", "_"): v for (k, v) in wordcloud.words_.items()}
wordcloud.generate_from_frequencies(underscore_separated)
```

**Comment by henriquebortoletti** [comment]
Yes, it works for any dictionary, but my point is about when you are counting the words, the text variable is getting the dictionary value for the text key without it being lower case, but in the second line, you are inserting in the dictionary the text key in lower case. 
`
val = tmpDict.get(text, 0)
`
`
tmpDict[text.lower()] = val + 1
`

**Comment by JoyLinWQ** [comment]
Hi @ricksilvab , @amueller , @Keramatfar , 

Any solutions so far? I've tried adding the `.lower()` as suggested above but to no avail. 
I have a slightly different scenario where i'm adding a list of stopwords to the predefined ones, but these words are still appearing in the wordcloud. Below is an example code for testing.

```
import matplotlib.pyplot as plt
import multidict as multidict
import re
from wordcloud import WordCloud, STOPWORDS

tokens_list = [
    'bigger', 'giraffe', 'one', 'on', 'left', 'top', 'giraffe', 'campbells', 'tomato', 'juice', 'can', 'campbells', 'label', 'guy', 'in', 'white', 'left', 'side', 'of', 'table', 'male', 'chef', 'man', 'on', 'left', 'middle', 'donut', 'far', 'right', 'far', 'right', 'glazed', 'bumpy', 'dount', 'donut', 'on', 'right', 'lying', 'on', 'top', 'white', 'cake', 'piece', 'blurry', 'the', 'cake', 'in', 'the', 'background', 'the', 'piece', 'of', 'cake', 'at', 'the', 'back', 'umpire', 'ump', 'umpire', 'person', 'with', 'ponytail', 'pony', 'tail', 'man', 'far', 'left', 'cathcer', 'catcgher', 'catcher', 'upper', 'far', 'right', 'banana', 'banana', 'top', 'right', 'banana', 'top', 'right', 'person', 'in', 'right', 'in', 'red', 'cap', 'guy', 'in', 'back', 'man', 'in', 'red', 'hard', 'hat', 'left', 'most', 'apple', 'slices', 'apple', 'chips', 'left', 'of', 'grapes', 'man', 'with', 'white', 'watch', 'looking', 'at', 'phone', 'middle', 'guy', 'guy', 'on', 'right', 'eating', 'rightmost', 'man', 'far', 'right', 'guy', 'looking', 'out', 'of', 'frame', 'front', 'bear', 'lower', 'bear', 'bear', 'closest', 'tan', 'couch', 'left', 'corner', 'bottom', 'left', 'couch', 'beige', 'couch', 'cushion', 'person', 'on', 'right', 'reaching', 'out', 'person', 'right', 'woman', 'in', 'purple', 'nearest', 'bed', 'closest', 'bed', 'bed', 'all', 'the', 'way', 'right', 'blue', 'bus', 'front', 'bus', 'blue', 'bus', 'bearded', 'guy', 'in', 'fez', 'guy', 'red', 'tie', 'center', 'man', 'in', 'fez', 'little', 'girl', 'on', 'the', 'right', 'leaning', 'over', 'little', 'girl', 'on', 'right', 'leaning', 'over', 'table', 'top', 'cup', 'top', 'cut', 'rose', 'girl', 'lady', 'with', 'game', 'controller', 'glasses', 'girl', 'red', 'blur', 'red', 'red', 'hat', 'left', 'side', 'food', 'at', 'back', 'back', 'dough', 'back', 'dough', 'right', 'couch', 'big', 'red', 'couch', 'couch', 'rainbow', 'umbrella', 'rainbow', 'umb', 'colorful', 'umbrella', 'left', 'couch', '3', 'cushion', 'couch', 'left', 'couch'
]


# do word cloud for tokens list based on word frequency counts
def getFrequencyDictForText(sentence):
    fullTermsDict = multidict.MultiDict()
    tmpDict = {}

    # making dict for counting frequencies
    for text in sentence.split(" "):
        if re.match("a|the|an|the|to|in|for|of|or|by|with|is|on|that|be", text):
            continue
        val = tmpDict.get(text.lower(), 0)      # add .lower()
        tmpDict[text.lower()] = val + 1
    for key in tmpDict:
        fullTermsDict.add(key, tmpDict[key])
    return fullTermsDict


def makeImage(text):

    stopwords = set(STOPWORDS)
    stopwords.update([
        "left", "right"
    ])
    
    wc = WordCloud(
        background_color="black", max_words=1000, max_font_size=100,
        stopwords=stopwords, random_state=42, collocations=False
    )
    # generate word cloud
    wc.generate_from_frequencies(text)

    # show
    plt.imshow(wc, interpolation="bilinear")
    plt.axis("off")
    plt.show()


text = " ".join(tokens_list).lower()
makeImage(getFrequencyDictForText(text))

```

With and without the change, I'm still getting the wordcloud attached where "left" and "right" are still appearing. 
<img width="498" alt="image" src="https://github.com/amueller/word_cloud/assets/69743962/b01fef73-7c78-4dfb-9f6e-f1c78b2c611a">


Please advise if the stopwords are working as intended. Thank you.
